syntax = "proto3";

package afs;

service FileSystemService 
{
    rpc Ping (PingMessage) returns (PingMessage) {}

    // File operations
    rpc Fetch (FetchRequest) returns (FetchResponse) {}
    rpc Store (StoreRequest) returns (StoreResponse) {}
    rpc Remove (RemoveRequest) returns (RemoveResponse) {}
    rpc Create (CreateRequest) returns (CreateResponse) {}
    // rpc Mknod (MknodRequest) returns (MknodResponse) {}
    
    // Metadata operations
    rpc GetFileStat (GetFileStatRequest) returns (GetFileStatResponse) {}
    rpc TestAuth (TestAuthRequest) returns (TestAuthResponse) {}
    rpc Access (AccessRequest) returns (AccessResponse) {}
    
    // Directory operations
    rpc MakeDir (MakeDirRequest) returns (MakeDirResponse) {}
    rpc RemoveDir (RemoveDirRequest) returns (RemoveDirResponse) {}
    rpc ReadDir (ListDirRequest) returns (ListDirResponse) {}

    rpc FetchUsingStream (FetchRequest) returns (stream FetchResponse) {}
    rpc StoreUsingStream (stream StoreRequest) returns (StoreResponse) {}
}

message PingMessage { }

message FetchRequest
{
    string pathname = 1;
}
message FetchResponse
{
    bytes file_contents = 1;
    Timestamp time_modified = 2;
    int32 fs_errno = 3;
}

message TestAuthRequest
{
    string pathname = 1;
    Timestamp time_modified = 2;
}
message TestAuthResponse
{
    bool file_changed = 1;
}

message GetFileStatRequest
{
    string pathname = 1;
}
message GetFileStatResponse
{
    FileStat stat = 1;
    int32 fs_errno = 2;
}

message AccessRequest
{
    string pathname = 1;
    int32 mode = 2;
}
message AccessResponse
{
    int32 fs_errno = 1;
}

// Based on chrono::timespec
// see https://en.cppreference.com/w/c/chrono/timespec
message Timestamp
{
    int64 sec = 1;
    uint32 nsec = 2;
}

enum FileMode {
    UNSUPPORTED = 0;
    REG = 1;
    DIR = 2;
}

message FileStat
{
    uint32   dev = 1;    
    uint32   ino = 2;
    int32   mode = 3;  
    uint32   nlink = 4;    
    uint32     uid = 5;     
    uint32     gid = 6;     
    sint64     size = 7;    
    sint64 blksize = 8; 
    sint64  blocks = 9;  
    int32    atime = 10;   
    int32    mtime = 11;   
    int32    ctime = 12; 
}

message DirectoryEntry {
    string file_name = 1;
    FileMode mode = 2;
    uint64 size = 3; // in bytes
}

message CreateRequest
{
    string pathname = 1;
    uint32 mode = 2;
    uint32 flags = 3;
}

message CreateResponse
{
    uint32 fs_errno = 1;
}

message StoreRequest
{
    string pathname = 1;
    bytes file_contents = 2;
}
message StoreResponse
{
    uint32 fs_errno = 1;
    Timestamp time_modify = 2;
}

message RemoveRequest 
{
    string pathname = 1;
}
message RemoveResponse
{
    uint32 fs_errno = 1;
}

message MknodRequest
{
    string pathname = 1;
    uint32 mode = 2;
    uint32 dev = 3;
}
message MknodResponse
{
    uint32 fs_errno = 1;
}

message MakeDirRequest
{
    string pathname = 1;
    int32 mode = 2;
}
message MakeDirResponse
{
    uint32 fs_errno = 1;
}

message RemoveDirRequest
{
    string pathname = 1;
}
message RemoveDirResponse
{
    uint32 fs_errno = 1;
}

message ListDirRequest
{
    string pathname = 1;
}
message ListDirResponse
{
    uint32 fs_errno = 1;
    repeated DirectoryEntry entries = 2;
}
